distributed/client.py

```diff
@@ -806,6 +808,150 @@ class VersionsDict(TypedDict):
     client: dict[str, dict[str, Any]]
 
 
+_T_LowLevelGraph: TypeAlias = dict[Key, tuple]
+
+
+class MapLayer(Layer):
+    # func: Callable
+    # iterable: Iterable
+    # key=None
+    # workers=None
+    # retries=None
+    # resources=None
+    # priority: int = 0
+    # allow_other_workers: bool = False
+    # fifo_timeout: str = "100 ms"
+    # actor: bool = False
+    # actors: bool = False
+    # pure: bool = True
+    # batch_size=None
+    # # **kwargs
+
+    def __init__(
+        self,
+        func: Callable,
+        iterables: Iterable,
+        key,
+        workers=None,
+        retries=None,
+        resources=None,
+        priority: int = 0,
+        allow_other_workers: bool = False,
+        fifo_timeout: str = "100 ms",
+        actor: bool = False,
+        actors: bool = False,
+        pure: bool = True,
+        batch_size=None,
+        parts_out=None,  # TODO check this
+        annotations=None,
+        **kwargs,
+    ):
+        self.func = func
+        self.iterables = iterables
+        self.key = key
+        self.workers = workers
+        self.retries = retries
+        self.resources = resources
+        self.priority = priority
+        self.allow_other_workers = allow_other_workers
+        self.fifo_timeout = fifo_timeout
+        self.actor = actor
+        self.actors = actors
+        self.pure = pure
+        self.batch_size = batch_size
+        self.kwargs = kwargs
+        self.parts_out = parts_out
+        super().__init__(annotations=annotations)
+
+    # TODO add repr
+    def __repr__(self) -> str:
+        raise NotImplementedError
+
+    @property
+    def _dict(self) -> _T_LowLevelGraph:
+        self._cached_dict: _T_LowLevelGraph
+        dsk: _T_LowLevelGraph
+
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict
+        else:
+            dsk = self._construct_graph()
+            self._cached_dict = dsk
+        return self._cached_dict
+
+    def get_output_keys(self) -> set[Key]:
+        return set(self.keys())
+        # return {(self.name, part) for part in self.parts_out}
+
+    def get_ordered_keys(self):
+        return list(self.keys())
+
+    def is_materialized(self) -> bool:
+        return hasattr(self, "_cached_dict")
+
+    def __getitem__(self, key: Key) -> tuple:
+        return self._dict[key]
+
+    def __iter__(self) -> Iterator[Key]:
+        return iter(self._dict)
+
+    def __len__(self) -> int:
+        return len(self._dict)
+
+    # TODO work this part out
+    # def _cull(self, parts_out: Iterable[int]) -> MapLayer:
```

Comment:
I think culling is not required. I'm not sure how to implement this to be ignored but the functionality is not necessary to be implemented